### ISSUE TITLE
Add example for dependency without pkg-config to the meson docs

### ISSUE THREAD

**Issue by textshell** [mesonbuild/meson#7281]
As far as i can see, currently the non reference part of the meson documentation does not have a description how to use libraries that don't provide pkg-config files (and are not one of the special cased libraries).
I think it would be good to add such information.

**Comment by thoni56** [comment]
Are there any tips on how to do this, given that it's not in the documentation?

**Comment by thoni56** [comment]
Ok, figured out that

    libcjson = meson.get_compiler('c').find_library('cjson')

and then adding libcjson to the link dependencies did work.

**Comment by eli-schwartz** [comment]
That actually will not work, because you will be unable to use the cjson headers.

For the cjson project you should use `dependency('libcjson')` (and optionally include the wrap file from https://wrapdb.mesonbuild.com/ as per the docs at https://mesonbuild.com/Wrap-dependency-system-manual.html)

**Comment by thoni56** [comment]
I know about `dependency()`. The problem is that `libcjson` 1.7.10 does not seem to have any `.pc` file in Ubuntu 20.04  since it uses 1.7.10-1  and the package files were added in 1.7.10-2 (http://changelogs.ubuntu.com/changelogs/pool/universe/c/cjson/cjson_1.7.13-1/changelog)

Looking up just the library worked fine in this case, The header files are installed in `/usr/include` so they are seen without explicit include paths. But I understand that this is pure chance.

So I guess for `libcjson1` in Ubuntu 20.04 we are back to "How do you, in the general case, configure the dependencies for a package that does not support pkg-config etc. nor is one of the special packages?"

(Wrap is when you want to "re-export" your dependencies to users of your library/package, right?)

**Comment by eli-schwartz** [comment]
> Looking up just the library worked fine in this case, The header files are installed in `/usr/include` so they are seen without explicit include paths. But I understand that this is pure chance.

No they are not installed in `/usr/include`.

They are installed in `/usr/include/cjson` and the pkg-config file adds that include path manually to the cflags, with the expectation that this makes `#include <cJSON.h>` work. But it will not work if you don't have an explicit include path.

That was my *entire* point.

...

Wrap is not for when you want to re-export your dependencies to users of your package.
Wrap is when you want to vendor your dependencies for users of your package.

Typically you would build with `meson setup -Dcjson:default_library=static` and install with `meson install --skip-subprojects`, although if your project exposes a library and also exposes cjson via its public headers (not its private headers! if cjson is merely an implementation detail of your project then its duty is entirely done once the cjson symbols are linked into your library) then people linking against your library will need cjson, in which case you may want to install subprojects. And you may also want to install to a private prefix to avoid clashes.

**Comment by eli-schwartz** [comment]
> we are back to "How do you, in the general case, configure the dependencies for a package that does not support pkg-config etc. nor is one of the special packages?"

There are long-term intentions to support adding some kind of mode where a dependency lookup can specify arbitrary fallbacks.

Currently in git master, you can do `dependency('thing', 'fallbackthing')` and meson will select the first dependency it finds in the list.
We'd eventually like to also be able to let you specify a fallback find_library+has_header, but there is no ETA on that...

**Comment by thoni56** [comment]
> > Looking up just the library worked fine in this case, The header files are installed in `/usr/include` so they are seen without explicit include paths. But I understand that this is pure chance.
> 
> No they are not installed in `/usr/include`.
> 
> They are installed in `/usr/include/cjson` and the pkg-config file adds that include path manually to the cflags, with the expectation that this makes `#include <cJSON.h>` work. But it will not work if you don't have an explicit include path.
> 
> That was my _entire_ point.

Yes, I think I got your point. I did actually mean to say "in a directory in /usr/include", which works because the software I'm trying to build does `#include <cjson/cJSON.h>`. Which again is "lucky". So I totally understand the dual "include+llb" thing `dependency()` does.
 
> ...
> 
> Wrap is not for when you want to re-export your dependencies to users of your package. Wrap is when you want to vendor your dependencies for users of your package.
> 
> Typically you would build with `meson setup -Dcjson:default_library=static` and install with `meson install --skip-subprojects`, although if your project exposes a library and also exposes cjson via its public headers (not its private headers! if cjson is merely an implementation detail of your project then its duty is entirely done once the cjson symbols are linked into your library) then people linking against your library will need cjson, in which case you may want to install subprojects. And you may also want to install to a private prefix to avoid clashes.

Thanks for that clarification.

> There are long-term intentions to support adding some kind of mode where a dependency lookup can specify arbitrary fallbacks.
>
>Currently in git master, you can do dependency('thing', 'fallbackthing') and meson will select the first dependency it finds in the list.
>We'd eventually like to also be able to let you specify a fallback find_library+has_header, but there is no ETA on that...

And I think this is what _my_ point was. So there is *no* way to get arbitrary header+libs without package discovery support into a meson build?

**Comment by eli-schwartz** [comment]
You can do

```
cc = meson.get_compiler('c')

foo_dep = dependency('foo')

# if pkg-config is not available, try heuristically grabbing libs
if not foo_dep.found()
    foo_dep = cc.find_library('foo', has_headers: ['foo.h'], required: true)
endif
```

You just cannot have it be an automatic fallback for `dependency('foo')`. You cannot use subproject wraps for the dependency while preferring the find_library. You cannot use feature options, since an enabled feature would immediately fail if the dependency is not found and never check the fallback find_library.

**Comment by thoni56** [comment]
That is an excellent example. Thank you!
